rust/src/mask/mod.rs

```diff
@@ -1,13 +1,204 @@
+//! - B4: all model weights are absolutely bounded by 10,000.
+//! - B6: all model weights are absolutely bounded by 1,000,000.
+//! - Bmax: all model weights are absolutely bounded by their primitive data type's absolute
+//!   maximum value.
+//!
+//! ## Model type
+//! The [`ModelType`] describes the maximum number of masked models that can be aggregated without
+//! information loss. The smaller the number of masked models, the less bytes are required to
+//! represent masked model weights. The model type variants are:
+//! - M3: at most 1,000 masked models may be aggregated.
+//! - M6: at most 1,000,000 masked models may be aggregated.
+//! - M9: at most 1,000,000,000 masked models may be aggregated.
+//! - M12: at most 1,000,000,000,000 masked models may be aggregated.
+//!
+//! # Masking, aggregation and unmasking
+//! Local models should be masked (i.e. encrypted) before they are communicated somewhere else to
+//! protect the possibly sensitive information learned from local data. The masking should allow
+//! for masked models to be aggregated while they are still masked (i.e. homomorphic encryption).
+//! Then the aggregated masked model can safely be unmasked without jeopardizing the secrecy of
+//! personal information if the model is generalized enough.
+//!
+//! ## Masking
+//! A [`Model`] can be masked with a [`Masker`], which requires a [`MaskConfig`]. During the
+//! masking the model weights are scaled, then embedded as elements of the chosen finite group and
+//! finally masked by randomly generated elements from that very same finite group. The scalar
+//! provides the necessary means to perform different aggregation strategies, for example federated
+//! averaging. The masked model is returned as a [`MaskObject`] and the mask used to mask the model
+//! can be generated via the additionally returned [`MaskSeed`].
+//!
+//! ```
+//! # use xain_fl::mask::{BoundType, DataType, FromPrimitives, GroupType, MaskConfig, Masker, Model, ModelType};
+//! // create local models and a fitting masking configuration
+//! let number_weights = 10;
+//! let scalar = 0.5;
+//! let local_model_1 = Model::from_primitives_bounded(vec![0_f32; number_weights].into_iter());
+//! let local_model_2 = Model::from_primitives_bounded(vec![1_f32; number_weights].into_iter());
+//! let config = MaskConfig {
+//!     group_type: GroupType::Prime,
+//!     data_type: DataType::F32,
+//!     bound_type: BoundType::B0,
+//!     model_type: ModelType::M3,
+//! };
+//!
+//! // mask the local models
+//! let (local_mask_seed_1, masked_local_model_1) = Masker::new(config).mask(scalar, local_model_1);
+//! let (local_mask_seed_2, masked_local_model_2) = Masker::new(config).mask(scalar, local_model_2);
+//!
+//! // derive the masks of the local masked models
+//! let local_mask_1 = local_mask_seed_1.derive_mask(number_weights, config);
+//! let local_mask_2 = local_mask_seed_2.derive_mask(number_weights, config);
+//! ```
+//!
+//! ## Aggregation
```

Comment:
sorry i should have used a single block for the below rather than chopping it up....